src/crush/CrushCompiler.cc

```diff
@@ -658,7 +663,25 @@ int CrushCompiler::parse_rule(iter_t const& i)
 	  err << "in rule '" << rname << "' item '" << item << "' not defined" << std::endl;
 	  return -1;
 	}
-	crush.set_rule_step_take(ruleno, step++, item_id[item]);
+        int c = -1;
+        string class_name;
+        if (s->children.size() > 2) {
+          class_name = string_node(s->children[3]);
+          if (!crush.class_rname.count(class_name)) {
+            err << "in rule '" << rname << "' class '" << class_name << "' not defined" << std::endl;
+            return -1;
+          }
+          c = crush.class_rname[class_name];
+	}
+        if (verbose) {
+          err << "rule " << rname << " take " << item;
+          if (c < 0)
+            err << std::endl;
+          else
+            err << " class " << class_name << std::endl;
+        }
+
+	crush.set_rule_step_take(ruleno, step++, item_id[item], c);
```

Comment:
I think ehre we actaully want to map item to the parallel per-class tree.

We could also pass c into the rule step since it's ignored, but i'd avoid doing that until we have a reason to use it within the rule processing itself..